File: src/seni/Focal.js
/*
 *  Seni
 *  Copyright (C) 2015 Inderjit Gill <email@indy.io>
 *
 *  This program is free software: you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation, either version 3 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program. If not, see <http://www.gnu.org/licenses/>.
 */

import PublicBinding from './PublicBinding';
import MathUtil from './MathUtil';
import Interp from './Interp';

function setupFocalParameters(publicBinding, params) {
  const fullParams = publicBinding.mergeWithDefaults(params);

  const {
    position,                   // the position of the point of interest
    distance,
    falloff
  } = fullParams;

  if (falloff !== 'linear' &&
      falloff !== 'quick' &&
      falloff !== 'slow-in' &&
      falloff !== 'slow-in-out') {
    console.log('invalid falloff value');
  }

  const fn = Interp.remapFn({
    from: [0, distance],
    to: [1, 0],
    mapping: falloff,
    clamping: true
  });

  return {
    position,
    fn
  };
}

function point(publicBinding, params, renderer) {

  const {
    position,
    fn
  } = setupFocalParameters(publicBinding, params);

  // returns a function that given a v2 returns how 'interesting' it should be
  return function(parameters) {
    const v = parameters.position || [0, 0];
    const p = renderer.vectorToCanvasSpace(v);
    const d = MathUtil.distance2d(position, p);

    return fn({val: d});
  };
}

const pointBinding = new PublicBinding(
  'focal/point',
  `returns a function that takes a v2 and returns the 'interest' at that point`,
  {
    position: [0, 0],
    distance: 100,
    falloff: 'linear'
  },
  (self, renderer) => params => point(self, params, renderer)
);


function vline(publicBinding, params, renderer) {
  const {
    position,
    fn
  } = setupFocalParameters(publicBinding, params);

  // returns a function that given a v2 returns how 'interesting' it should be
  return function(parameters) {
    const v = parameters.position || [0, 0];
    const p = renderer.vectorToCanvasSpace(v);
    const d = MathUtil.distance1d(position[0], p[0]);

    return fn({val: d});
  };
}

const vlineBinding = new PublicBinding(
  'focal/vline',
  `returns a function that takes a v2 and returns the 'interest' at that point`,
  {
    position: [500, 500],
    distance: 100,
    falloff: 'linear'
  },
  (self, renderer) => params => vline(self, params, renderer)
);


function hline(publicBinding, params, renderer) {
  const {
    position,
    fn
  } = setupFocalParameters(publicBinding, params);

  // returns a function that given a v2 returns how 'interesting' it should be
  return parameters => {
    const v = parameters.position || [0, 0];
    const p = renderer.vectorToCanvasSpace(v);
    const d = MathUtil.distance1d(position[1], p[1]);

    return fn({val: d});
  };
}

const hlineBinding = new PublicBinding(
  'focal/hline',
  `returns a function that takes a v2 and returns the 'interest' at that point`,
  {
    position: [500, 500],
    distance: 100,
    falloff: 'linear'
  },
  (self, renderer) => params => hline(self, params, renderer)
);

const Focal = {
  publicBindings: [
    pointBinding,
    vlineBinding,
    hlineBinding
  ]
};

export default Focal;

File: core/src/focal.rs
// Copyright (C) 2020 Inderjit Gill <email@indy.io>

// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU Affero General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.

// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Affero General Public License for more details.

// You should have received a copy of the GNU Affero General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.

use crate::context::Context;
use crate::ease::Easing;
use crate::interp;
use crate::mathutil;

#[derive(Copy, Clone, Debug)]
pub enum FocalType {
    Point,
    HLine,
    VLine,
}

#[derive(Clone, Debug)]
pub struct FocalStateStruct {
    pub focal_type: FocalType,
    pub mapping: Easing,
    pub position: (f32, f32),
    pub distance: f32,
    pub transform_pos: bool,
}

impl Default for FocalStateStruct {
    fn default() -> FocalStateStruct {
        FocalStateStruct {
            focal_type: FocalType::Point,
            mapping: Easing::Linear,
            position: (0.0, 0.0),
            distance: 1.0,
            transform_pos: false,
        }
    }
}

impl FocalStateStruct {
    pub fn value(&self, context: &Context, position: (f32, f32)) -> f32 {
        // transform position to canvas space coordinates
        let (x, y) = if self.transform_pos {
            context.matrix_stack.transform_vec2(position.0, position.1)
        } else {
            position
        };

        match self.focal_type {
            FocalType::Point => point(x, y, self.distance, self.mapping, self.position),
            FocalType::HLine => hline(y, self.distance, self.mapping, self.position.1),
            FocalType::VLine => vline(x, self.distance, self.mapping, self.position.0),
        }
    }
}

fn point(x: f32, y: f32, distance: f32, mapping: Easing, centre: (f32, f32)) -> f32 {
    let d = mathutil::distance_v2(x, y, centre.0, centre.1);

    if d < std::f32::EPSILON {
        return 1.0;
    }

    interp::parametric(d, 0.0, distance, 1.0, 0.0, mapping, true)
}

fn hline(y: f32, distance: f32, mapping: Easing, centre_y: f32) -> f32 {
    let d = (centre_y - y).abs();

    if d < std::f32::EPSILON {
        return 1.0;
    }

    interp::parametric(d, 0.0, distance, 1.0, 0.0, mapping, true)
}

fn vline(x: f32, distance: f32, mapping: Easing, centre_x: f32) -> f32 {
    let d = (centre_x - x).abs();

    if d < std::f32::EPSILON {
        return 1.0;
    }

    interp::parametric(d, 0.0, distance, 1.0, 0.0, mapping, true)
}
